prov/efa/src/efa_user_info.c

```diff
@@ -269,34 +269,133 @@ int efa_user_info_get_dgram(uint32_t version, const char *node, const char *serv
+	 * of FI_HMEM as long as CUDA library is initialized and a CUDA device is
+	 * available. On an system without GPUDirect RDMA, the support of CUDA memory
+	 * is implemented by calling CUDA library. If user does not want EFA provider
+	 * to use CUDA library, the user can call use fi_setopt to set
+	 * FI_OPT_CUDA_API_PERMITTED to false. On an system without GPUDirect RDMA,
+	 * the call would fail.
+	 */
+	if (FI_VERSION_GE(version, FI_VERSION(1, 18)) && hmem_ops[FI_HMEM_CUDA].initialized) {
+		EFA_INFO(FI_LOG_CORE,
+			"User is using 1.18 API and CUDA library and device are availble, cliam support of HMEME.");
+		return true;
+	}
+
+	if (ofi_hmem_p2p_disabled()) {
+		EFA_WARN(FI_LOG_CORE,
+			"FI_HMEM capability currently requires peer to peer "
+			"support, which is disabled because FI_HMEM_P2P_DISABLED was set to 1/on/true.");
+		return false;
+	}
+
+	if (!efa_device_support_rdma_read()) {
+		EFA_WARN(FI_LOG_CORE,
+			"FI_HMEM capability requires RDMA, which this device does not support.");
+		return false;
+	}
+
+	if (!rxr_env_get_use_device_rdma()) {
+		EFA_WARN(FI_LOG_CORE,
+			"FI_HMEM capability requires RDMA, which is turned off. "
+			"You can turn it on by setting environment variable "
+			"FI_EFA_USE_DEVICE_RDMA to 1.\n");
+		return false;
+	}
+
+	return true;
+}
+
+#else
+
+bool efa_user_info_should_support_hmem(int version, const struct fi_info *hints)
+{
+	return false;
+}
+
+#endif
 /**
  * @brief update an info to match user hints
  *
  * the input info is a duplicate of prov info, which matches
  * the capability of the EFA device. This function tailor it
  * so it matches user provided hints
  *
+ * @param	version[in]	libfabric API version
  * @param	info[in,out]	info to be updated
  * @param	hints[in]	user provided hints
  * @return	0 on success
  * 		negative libfabric error code on failure
  */
 static
-int efa_user_info_alter_rxr(struct fi_info *info, const struct fi_info *hints)
+int efa_user_info_alter_rxr(int version, struct fi_info *info, const struct fi_info *hints)
 {
 	uint64_t atomic_ordering;
-	bool support_atomic;
-
-	/*
-	 * Do not advertise FI_HMEM capabilities when the core can not support
-	 * it or when the application passes NULL hints (given this is a primary
-	 * cap). The logic for device-specific checks pertaining to HMEM comes
-	 * further along this path.
-	 */
-	if (!hints) {
+	if (efa_user_info_should_support_hmem(version, hints)) {
+		info->caps |= FI_HMEM;
+	} else {
 		info->caps &= ~FI_HMEM;
 	}
 
+	if (info->caps & FI_HMEM) {
+		/* Add FI_MR_HMEM to mr_mode when claiming support of FI_HMEM
+		 * because EFA provider's HMEM support rely on
+		 * application to provide descriptor for device buffer.
+		 * Note we did not add FI_MR_LOCAL here because according to FI_MR man
+		 * page:
+		 *
+		 *     "If FI_MR_HMEM is set, but FI_MR_LOCAL is unset,
+		 *      only device buffers must be registered when used locally.
+		 *      "
```

Comment:
Suggest removing this quote. The man page might change later, and it is unnecessary work to keep this in sync.

Reply:
removed in in new revision.